go/libraries/doltcore/sqle/dsess/session.go

```diff
@@ -570,16 +570,17 @@ func (d *DoltSession) NewPendingCommit(ctx *sql.Context, dbName string, roots do
 			mergeParentCommits = append(mergeParentCommits, parentCommit)
 		}
 
-		err = actions.ResetSoftToRef(ctx, sessionState.dbData, "HEAD~1")
+		root, err := actions.ResetSoftToRef(ctx, sessionState.dbData, "HEAD~1")
```

Comment:
It would be better if we didn't do this... the problem is that the action calls repoStateWriter.UpdateStagedRoot, which in turn calls session.UpdateWorkingSet, which 1) has its own update semantics / lifecycle that interact with the work being done here, and 2) are inappropriate in the context of what *should* be a purely functional constructor

A better solution would be for the action to return a doiltdb.Roots struct, which the caller can decide how to apply to update state as necessary. Since this is just on the --amend path this isn't critical, maybe add a TODO